mlflow/store/dbmodels/models.py

```diff
@@ -0,0 +1,224 @@
+    except ImportError:
+        return 'Unknown'
+
+
+ExperimentLifecycleStages = [
+    Experiment.ACTIVE_LIFECYCLE,
+    Experiment.DELETED_LIFECYCLE
+]
+
+SourceTypes = [
+    SourceType.NOTEBOOK,
+    SourceType.JOB,
+    SourceType.LOCAL,
+    SourceType.UNKNOWN,
+    SourceType.PROJECT
+]
+
+RunStatusTypes = [
+    RunStatus.SCHEDULED,
+    RunStatus.FAILED,
+    RunStatus.FINISHED,
+    RunStatus.RUNNING
+]
+
+LifecycleStageTypes = [
+    RunInfo.ACTIVE_LIFECYCLE,
+    RunInfo.DELETED_LIFECYCLE
+]
+
+
+def generate_uuid():
+    return uuid.uuid4().hex
+
+
+def _validate(self):
+    if not hasattr(self, '__entity__') and self.__entity__ is not None:
+        raise Exception(
+            'sqlalchemy model <{}> needs __entity__ set'.format(self.__class__.__name__))
+
+    if not hasattr(self, '__properties__') and self.__entity__ is not None:
+        raise Exception(
+            'sqlalchemy model <{}> needs __properties__ set'.format(self.__class__.__name__))
+
+
+def _to_mlflow_entity(self):
+    _validate(self)
+
+    # create dict of kwargs properties for entity and return the intialized entity
+    config = {}
+    for k in self.__properties__:
+        # check if its mlflow entity and build it
+        obj = getattr(self, k)
+        try:
+            config[k] = obj.to_mlflow_entity()
+        except AttributeError:
+            if k in ['metrics', 'params', 'tags']:
+                # these are list so
+                obj = [v.to_mlflow_entity() for v in obj]
+
+            config[k] = obj
+
+    return self.__entity__(**config)
+
+
+class SqlExperiment(Base):
+    __tablename__ = 'experiments'
+    __entity__ = Experiment
+    __properties__ = Experiment._properties()
+    experiment_id = Column(Integer, primary_key=True)
+    name = Column(String(256), unique=True, nullable=False)
+    artifact_location = Column(Text, nullable=True)
+    lifecycle_stage = Column(Integer, default=Experiment.ACTIVE_LIFECYCLE)
+
+    __table_args__ = (
+        CheckConstraint(
+            lifecycle_stage.in_(ExperimentLifecycleStages), name='lifecycle_stage'),
+    )
+
+    def __repr__(self):
+        return '<SqlExperiment ({}, {})>'.format(self.experiment_id, self.name)
+
+    def to_mlflow_entity(self):
+        return _to_mlflow_entity(self)
+
+
+class SqlRunTag(Base):
+    __tablename__ = 'run_tag'
+    __entity__ = RunTag
+    __properties__ = RunTag._properties()
+    id = Column(Integer, primary_key=True)
+    run_data_id = Column(Integer, ForeignKey('run_data.id'))
+    run_data = relationship('SqlRunData', backref='tags')
+    key = Column(Text, nullable=False)
+    value = Column(Text, nullable=True)
+
+    def __repr__(self):
+        return '<SqlRunTag({}, {})>'.format(self.key, self.value)
+
+    def to_mlflow_entity(self):
+        return _to_mlflow_entity(self)
+
+
+class SqlMetric(Base):
+    __tablename__ = 'metric'
+    __entity__ = Metric
+    __properties__ = Metric._properties()
+    id = Column(Integer, primary_key=True)
```

Comment:
Is this primary key generated? Conceptually--metrics, tags, and parameters do not have an ID. As implemented in `FileStore`, uniqueness is defined in other ways. 


|entity|unique combination|
|---|---|
|metrics|(experiment_id, run_id, name, timestamp)|
|params|(experiment_id, run_id, name, value)|
|tag|(experiment_id, run_id, name, value)|

For `FileStore` implementation that is done in python code. For SQL store can this enforcement be pushed down to the implementation?

Reply:
The id columns are just primary keys we need for sql itself they don’t get used. The metrics, tags, params show as lists in SqlRun model so that we can do (will rewrite it to be) Run.metrics or run.metrics[0]  so each run model has its own list of metrics joined by that id

Reply:
How would it be forced down to implementation?